windows cross compile

hamishmack opened the pull request:
Just getting hydra started.

michaelpj commented on default.nix:
```diff
@@ -31,6 +31,7 @@ let
   inherit (pkgs) lib haskell-nix;
   inherit (plutus) haskell agdaPackages;
   inherit (plutus) easyPS sphinxcontrib-haddock;
+  noCross = x: if crossSystem == null then x else {};
```
Isn't this handled by the guards in `ci.nix` already?

hamishmack replied:
switched to using `meta.platforms`

michaelpj commented on nix/tests/default.nix:
```diff
@@ -19,14 +19,15 @@
   shellcheck = pkgs.callPackage ./shellcheck.nix { src = cleanSrc; };
 
-  stylishHaskell = pkgs.callPackage ./stylish-haskell.nix {
+  stylishHaskell = noCross (pkgs.callPackage ./stylish-haskell.nix {
```
Same question: let's just not ask for `tests` in ci.nix when we're cross.

michaelpj replied:
I still think this might be simpler? Or is the issue that it won't even evaluate? But that ought to also be okay if we just didn't pull in the test attrset if we're cross...

michaelpj reviewed (commented):
To be clear, we really only care about `plutus-ledger-api` cross-compiling right now. In particular, it's fine if packages that use the plugin don't work!

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -138,7 +193,7 @@ let
           plutus-pab.ghcOptions = [ "-Werror" ];
           plutus-tx.ghcOptions = [ "-Werror" ];
-          plutus-tx-plugin.ghcOptions = [ "-Werror" ];
+          plutus-tx-plugin.ghcOptions = lib.optional (!pkgs.stdenv.hostPlatform.isWindows) "-Werror";
```
this should probably be a mkIf?

hamishmack replied:
this was not needed in the end

michaelpj commented:
Very weird eval error now:
```
error: [json.exception.type_error.302] type must be string, but is null
```

johnalotoski commented:
This pr has been failing eval on hydra since 2021-06-30.  Disabling automatic eval and build with the `hydra-dont-build` label to avoid unnecessary hydra resource consumption.  Please remove this label from the PR when you want hydra to resume eval and builds.

michaelpj commented on fake-pab/default.nix:
```diff
@@ -1,14 +1,15 @@
 { pkgs, gitignore-nix, haskell, webCommon, webCommonMarlowe, buildPursPackage, buildNodeModules, filterNpm, marlowe-dashboard }:
 let
+  nativeOnly = pkgs.lib.meta.addMetaAttrs { platforms = with pkgs.lib.platforms; [ linux darwin ]; };
```
We define this in a few places, should we put it somewhere common and pass it in?

michaelpj commented on nix/docs.nix:
```diff
@@ -48,31 +49,31 @@ let
-    eutxo = import ../papers/eutxo { inherit buildLatexDoc; };
-    utxoma = import ../papers/utxoma { inherit buildLatexDoc; };
-    eutxoma = import ../papers/eutxoma { inherit buildLatexDoc; };
+    system-f-in-agda = nativeOnly (import ../papers/system-f-in-agda { inherit buildLatexDoc; });
```
maybe just make the latex builder set the meta attributes? I think it would be fine to say that all docs only build natively by default!

michaelpj commented on nix/docs.nix:
```diff
@@ -48,31 +49,31 @@ let
+  cost-model-notes = nativeOnly (import ../notes/cost-model-notes { inherit buildLatexDoc; });
 
-  site = pkgs.callPackage ../doc {
+  site = nativeOnly (pkgs.callPackage ../doc {
```
just add the meta attributes in the derivation?

michaelpj commented on nix/pkgs/default.nix:
```diff
@@ -181,10 +182,14 @@ let
-    };
+    pkgs.lib.addMetaAttrs
+      {
+        platforms = with pkgs.lib.platforms; [ linux darwin ];
```
just add them in the derivation?

michaelpj commented on nix/pkgs/haskell/extra.nix:
```diff
@@ -18,7 +18,11 @@ let
+    plan-sha256 = lib.removeSuffix "\n" (builtins.readFile (
+      if stdenv.hostPlatform.isWindows
+      then ./agda-windows.sha
+      else ./agda.sha
```
:see_no_evil:

shlevy replied:
:scream:

michaelpj commented on nix/pkgs/haskell/extra.nix:
```diff
@@ -98,6 +102,7 @@ let
   updateAllShaFiles = writeShellScript "updateShaFiles" ''
     ${updateShaFile cabalInstallProject ./cabal-install.sha}
     ${updateShaFile agdaProject ./agda.sha}
+    ${updateShaFile agdaProject.projectCross.mingwW64 ./agda-windows.sha}
```
Shouldn't this just happen by us running this with cross platforms in `updateMaterialized`?

hamishmack replied:
A checked and without this line running:
```
nix-shell --run updateMaterialized
```
does not update `agda-windows.sha`

michaelpj commented on plutus-pab-client/default.nix:
```diff
@@ -1,10 +1,11 @@
   test-generator = haskell.packages.plutus-pab.components.exes.plutus-pab-test-psgenerator;
 
-  generated-purescript = pkgs.runCommand "plutus-pab-purescript" { } ''
+  generated-purescript = pkgs.runCommand "plutus-pab-purescript" { meta.platforms = with pkgs.lib.platforms; [ linux darwin ]; } ''
```
nativeOnly?

michaelpj replied:
There's a bit of inconsistency with how you handle the `runCommand` instances.

Also, I do think for stuff like this where there's a whole tranche of stuff that won't work in cross, it might be simpler to just not include them when we're cross.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -57,9 +56,93 @@ let
       "https://github.com/input-output-hk/Win32-network"."94153b676617f8f33abe8d8182c37377d2784bd1" = "0pb7bg0936fldaa5r08nqbxvi2g8pcy4w3c7kdcg7pdgmimr30ss";
       "https://github.com/input-output-hk/hedgehog-extras"."8bcd3c9dc22cc44f9fcfe161f4638a384fc7a187" = "12viwpahjdfvlqpnzdgjp40nw31rvyznnab1hml9afpaxd6ixh70";
     };
+    cabalProjectLocal = lib.optionalString pkgs.stdenv.hostPlatform.isWindows ''
```
Is there a reason not to do these things with module option settings?

hamishmack replied:
`cabalProjectLocal` is used by `cabal configure`, module option settings are only looked at later.  IIRC these were all things to help make `cabal configure` work when cross compiling for windows.

michaelpj replied:
Ugh I see, so we need to do this if we need the options at configure time and can't set them later. I feel like something has gone wrong if we need to do that, but I don't know what else we could do. Worth a comment, anyway.

michaelpj replied:
Should the conditional here be a generic cross conditional? i.e. hostPlatform != buildPlatform?

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -57,9 +56,93 @@ let
+      package plutus-tx-plugin
+        flags: +use-ghc-stub
+
+      -- The following is needed because Nix is doing something crazy.
```
Do we need these?

hamishmack replied:
Without `use-ghc-stub` I think the solver will fail while trying to find a `ghc` package.  I'll double check and update the comment to reflect why they are needed.

michaelpj replied:
Still needs a better comment

michaelpj replied:
bump

hamishmack replied:
Just two of these were needed and I have added a comment explaining that it is the `doctest` dependency that causes issues.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -157,13 +251,30 @@ let
           cardano-crypto-praos.components.library.pkgconfig = lib.mkForce [ [ libsodium-vrf ] ];
           cardano-crypto-class.components.library.pkgconfig = lib.mkForce [ [ libsodium-vrf ] ];
+
+          # By default haskell.nix chooses the `buildPackages` versions of these `build-tool-depeends`, but
```
Is this a haskell.nix bug?

hamishmack replied:
I think it is because we are using `build-tool-depends` as if it was `run-tool-depends`.

hamishmack replied:
So haskell.nix is interpreting it correctly.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -122,6 +203,19 @@ let
+            platforms = lib.platforms.linux;
+          };
+
+          # Windows build of libpq is marked as broken
```
Would putting the `if os(windows) buildable: false` conditional in cabal files fix these also?

hamishmack replied:
libpq is only broken for windows cross compile with nix, but it might build on windows itself.

michaelpj replied:
Right, but what you're putting here is *also* saying it will only build on linux or darwin. So while it may be overly zealous to say that `fake-pab` etc won't build on windows, I don't think it's too bad.

hamishmack replied:
I guess I felt the restriction belonged in nix because libpq was likely only broken in nix.  We can put it in the `.cabal` file, but it will prevent someone building `fake-pab` etc. on windows (even if they have working libpq installed).  Is `fake-pab` something `cardano-node` might use for testing on their windows github builders in the future?

michaelpj replied:
Right now, the only library that we actually care about *any* kind of windows support for is `plutus-ledger-api`. That's the thing that `cardano-node` depends on. I know people will be cross-compiling that and maybe even native-compiling it.

Everything else is gravy right now. Maybe later we'll need to revisit this, but for now I'm mostly in favour of doing the simplest and most maintainable thing.

michaelpj replied:
still relevant

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -57,9 +56,93 @@ let
+          ];
+        packages.Win32.components.library.build-tools = lib.mkForce [ ];
+        # TODO fix nix-tools so this is automatic
+        packages.marlowe.components.tests.marlowe-test.buildable = lib.mkForce false;
```
ditto for these?

michaelpj replied:
Or indeed I don't know why this isn't handled by either setting platforms or a buildable conditional in the cabal file?

michaelpj commented:
Mostly looking good, the main things are:
- Setting platforms on things vs. just not including them in `ci.nix` based on whether we're cross.
    - I think for e.g. the haskell packages, it's easier to include them all in `ci.nix` and then filter based on platforms. But for stuff like the tests or the docs attrset we just don't want *any* of them, so we might as well not include the attrset.
- The many and varied ways of saying "don't build this on windows/cross".
    - It seems like there are 3-4 ways of doing this - I am confused!

hamishmack commented:
I thought "Isn't this handled by the guards in ci.nix already?"  https://github.com/input-output-hk/plutus/pull/3430#discussion_r658614215

That by `guards` you meant this: https://github.com/input-output-hk/plutus/blob/636e1df540f3e8ee54ae93670949b74079e28e47/ci.nix#L59

I'll change `ci.nix` to exclude tests entirely for windows.

michaelpj commented:
So we have two ways of controlling what gets built in `ci.nix`:
- Setting platforms on the derivations, and allowing the filter to filter them out (e.g. all the `nativeOnly` stuff)
- Using conditionals in `ci.nix` itself to not include things (e.g. https://github.com/input-output-hk/plutus/pull/3430/files#diff-783b1f6267f9b548283102092f66f71ac066b9e7b0917de0663e0269e1e26848R73)

It's not 100% clear that one of these is better than the other in all cases. I was trying to articulate a heuristic like:
- If we have an attribute set where various members of it will or won't be available on different platforms, then try and set the platforms for derivations appropriately, include the set, and let the filter work.
- If we have an attribute set where we just plain know we don't care about building it on most platforms, then just only include it where we want it, and don't fuss around with the platforms.

We already have a couple of things in category 2, namely the shell and the deployment stuff. I think we could safely add basically everything on this line too: https://github.com/input-output-hk/plutus/pull/3430/files#diff-783b1f6267f9b548283102092f66f71ac066b9e7b0917de0663e0269e1e26848R69

I think that would then let us ditch most of the `nativeOnly` stuff.

michaelpj commented:
I guess maybe at some point I had a fantasy of just controlling everything with `meta.platforms`, but honestly it's kind of tedious doing it thoroughly, especially for stuff where we know we are only interested in it in some cases.

hamishmack commented on cabal.project:
```diff
@@ -27,6 +27,7 @@ packages: doc
           quickcheck-dynamic
           web-ghc
           word-array
+          stubs/plutus-ghc-stub
```
Done

hamishmack commented:
I changed `ci.nix` so it excludes jobs except those in `haskell` when cross compiling.  I have moved all the code setting `platforms` to another branch on my local repo as it is no longer needed.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -58,9 +57,96 @@ let
+            "xhtml"
+            # "stm" "terminfo"
+          ];
+        packages.Win32.components.library.build-tools = lib.mkForce [ ];
```
mysterious

michaelpj replied:
bump

hamishmack replied:
I think this was only needed when we had more build time dependencies.  I have removed it for now. Cabal is picking the Win32 that comes with ghc 8.10 and if that becomes an issue (as it has in some other projects) and we request a new version, we might need to add this back.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -58,9 +57,96 @@ let
+            # "stm" "terminfo"
+          ];
+        packages.Win32.components.library.build-tools = lib.mkForce [ ];
+        # TODO fix nix-tools so this is automatic
```
mysterious

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -58,9 +57,96 @@ let
+        # Remove hsc2hs build-tool dependencies (suitable version will be available as part of the ghc derivation)
+        packages.terminal-size.components.library.build-tools = lib.mkForce [ ];
+        packages.network.components.library.build-tools = lib.mkForce [ ];
+        packages.process.components.library.libs = lib.mkForce [ ];
```
what's this one?

hamishmack replied:
This is gone now too.  These packages include a `build-tool-depends` on `hsc2hs`.  IIRC the plan we generated caused us to try to build hsc2hs, but with the wrong Win32.  Just telling haskell.nix not to try to build it worked.

michaelpj commented on release.nix:
```diff
@@ -6,11 +6,17 @@
 , checkMaterialization ? false
 }:
 let
+  traceNames = prefix: builtins.mapAttrs (n: v:
```
probably don't want to commit this? not sure

hamishmack replied:
I have found it very useful when debugging hydra eval errors.  Some times you don't know what it was evaluating when the error happened (since the hydra log does not include the `stdout` output of `hydra-eval-jobs`).  With this you can tell what you need to try with `nix-build release.nix -A` or what to look for in `nix repl release.nix` to reproduce the error.

michaelpj replied:
Fine with me!

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -148,36 +148,26 @@ let
-	    packages.plutus-metatheory.components.library.buildable = lib.mkForce false;
+        packages = {
+          # Things that need plutus-tx-plugins
+          marlowe.package.buildable = false;
```
Now *here's* an example of where a cabal `buildable` (or `isCross`?) flag would make life easier, since we already have the ability to set flags on a per-package basis!

hamishmack replied:
We should revisit this, but for now I think excluding them in `nix` makes the most sense.  It is kind of `haskell.nix` that is at fault to provide a cross compiler with support for plugins.

If we did do it with a cabal flag I wonder if `no-plutus-tx-plugin` a nicer flag to have.  It might be a pain to maintain that in all the `.cabal` files though.

shlevy reviewed (approved):
Modulo @michaelpj's open comments, this looks good to me

michaelpj commented:
Actually, on the agda topic... when we're cross, could we just disable the stuff that depends on Agda? Only `plutus-metatheory` does, which we don't care about cross-compiling. Then we can not worry about the sha...

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+    # to windows. We can't use `modules` for these as `modules` are only applied
+    # after cabal has been configured.
+    cabalProjectLocal = lib.optionalString pkgs.stdenv.hostPlatform.isWindows ''
+      -- When cross compiling we don't have a `ghc` package
```
the conditional is "building for windows" not "cross-building" - which one should it be? Presumably if you were doing a native build on windows this would be okay?

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+        tests: False
+
+      package cardano-node
+        flags: -systemd
```
on the other hand, this stuff seems like it probably *is* related to building for windows full stop

michaelpj replied:
Actually, do we even need this line any more given https://github.com/input-output-hk/cardano-node/blob/master/cardano-node/cardano-node.cabal#L55 ?

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+      })
+      ({ pkgs, ... }: lib.mkIf (pkgs.stdenv.hostPlatform != pkgs.stdenv.buildPlatform) {
+        packages = {
+          # Things that need plutus-tx-plugins
```
```suggestion
          # Things that need plutus-tx-plugin
```

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+          # Needs agda
+          plutus-metatheory.package.buildable = false;
+          # These need R
+          plutus-core.components.benchmarks.cost-model-test.buildable = lib.mkForce false;
```
why the mkForce?

hamishmack replied:
It's because we do not use `mkDefault` when setting the `buildable` expression in `cabal-to-nix`.

hamishmack replied:
I think we spoke before about changing stuff in `cabal-to-nix` to use `mkDefault` so that we don't need mkForce as much.  I think any option that `cabal-to-nix` always sets should probably use `mkDefault` and I think `buildable` is always set.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+      })
+      ({ pkgs, ... }:
+        let
+          symlinkDlls = ''
```
needs a comment. Where does this come from? What is it for? Can we fix this in haskell.nix?

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+            ln -s ${pkgs.windows.mcfgthreads}/bin/mcfgthread-12.dll $out/bin/mcfgthread-12.dll
+          '';
+        in
+        lib.mkIf (pkgs.stdenv.hostPlatform.isWindows) {
```
can we merge this with the other `hostPlatform.isWindows` module? So we have one windows module, one cross module, and one unconditional module

hamishmack replied:
Turns out we don't need the module to make Win32 reinstallable any more.

michaelpj commented on nix/pkgs/haskell/haskell.nix:
```diff
@@ -59,9 +58,138 @@ let
+        flags: -systemd
+
+      package cardano-config
+        flags: -systemd
```
this is actually totally redundant, this flag does nothing on that package

michaelpj reviewed (approved):
LGTM! Thanks for all the work on this!
